### ISSUE TITLE
Prettier (+/- colors for semantics) and more readable codeblocks

### ISSUE THREAD

**Issue by aartaka** [atlas-engineer/nyxt#3192]
**Purpose of request:**
Our codeblocks are quire complex, packing symbol references, code, and formatting. One problem with them is theming: built-in operators are colored in `action` color, while referenciable symbols are colored `primary`. The problem is: `action` and `primary` are not guaranteed to work well with `codeblock` color.

**Describe your proposed change:**
I'm thinking of using `codeblock-` for built-in symbols and `codeblock+` for referenced ones. And use these colors are _background_ colors, so that it doesn't break on `codeblock` color.

**Why do you believe this to be an improvement?**
Code blocks become more theme-friendly and, as a consequence, more readable.

**Additional context/User story:**
See the [Modus theme screenshot](https://github.com/atlas-engineer/nyxt/blob/master/libraries/theme/custom-modus-theme.png) in the `theme` README: it's somewhat messed up.

**Screenshots/Mock ups:**
Throwing that from the top of my head. If I get to implementing it, I'll probably mock it up, share the screenshot, and we can decide whether we should have it.

---

### Pre-Build Checklist:
- [ ] Discussion has occurred.
- [ ] One or more developers have signed off on this change.
- [ ] One or more user researchers have signed off on this change.

**Comment by aartaka** [comment]
> I think that the only important aspect is that Nyxt's default themes need to be made properly. Custom themes are the the author's responsibility and they may not follow the contrast directives we follow.

No. The whole point of theming is that we provide a set of APIs conforming to which one can make a pretty theme equal to built-in ones. Yes, if someone injects custom CSS extending the theme beyond our APIs—it's their own problem. But all the otherwise, all themes should be born equal.

And note that it's not (only) about contrast anymore: contrast is ensured automatically. It's about semantics of themes and making themes understandable and reproducible by anyone.

> Generally speaking, any reference to external projects is not desirable (including emacs, vi, etc).

A counter-example: we mention Google, Pypi, DOI, DuckDuckGo already. Quite explicitly, in the manual. How is Modus theme different? Note that Solarized (!) is a project too. And Darkula. And Monokai. Most themes are separate external projects, so you can't ~make an omelette without breaking eggs~ make theming without mentioning themes.

> And it's not the Modus themes that came up with the idea of ensuring contrast ratios for readability.

I'm not mentioning that anywhere.

**Comment by aartaka** [comment]
And then, this issue also influences built-in themes, and will likely influence themes that @lansingthomas is baking. So we have to address it universally, whatever the support status for external themes.

**Comment by lansingthomas** [comment]
Would more text colors help?

Something like:
text: code-alt
text: code-symbol
text: code-extra
text: code-fun
text: code-parens

It is kind of cool to be able to set your own code colors

**Comment by aartaka** [comment]
>Would more text colors help?
>
>Something like:
>text: code-alt
>text: code-symbol
>text: code-extra
>text: code-fun
>text: code-parens


Sounds like text+ and text- to me. Maybe use these (+/-) text variations instead? Requires that codeblock text is in sync with the background one, though...

**Comment by lansingthomas** [comment]
I don't think I understand the proposal.

What parts am I missing?

Let me try to list the elements of a code block in default state in Nyxt, and @aartaka maybe you can fill in the missing parts.

`codeblock text`;
- text color {`text` from the text colors palette}
- text size and style {from the typography guideline}
- background color {`codeblock` from surface colors palette}
- background size and shape {we have not established an explicit convention for this yet}
- ????? help

**Comment by lansingthomas** [comment]
The more I reflect on this the more I want to try some more colors for the built in operators and the symbols. Our palette is SOOOO minimal already. 

@aartaka can you send me a nice code example snippet with all the elements?

**Comment by aartaka** [comment]
Here's the same code block in both themes. Legend:
- Primary link (`define-configuration`): macro, special syntax.
- Regular link (`make-keymap`): function, class, or any other "normal" program entity.
- Action color, no link (`let`): special built-in form.

![code-entities-light](https://github.com/atlas-engineer/nyxt/assets/57838654/6ddcfe63-68e7-4a2a-aaf5-d5178b1fc48e)
![code-entities-dark](https://github.com/atlas-engineer/nyxt/assets/57838654/1df45896-125c-44d1-bb97-5f403310a0c2)

(Wow, I haven't ran latest Nyxt for some time, light theme codeblock color looks SLEEEEEK :star_struck:)

**Comment by lansingthomas** [comment]
record of a conversation:

Artyom and Tom spoke about code block styling, and considered three options moving forward to help differentiate typographic code elements like `special forms`, `code macros`, ... perhaps even things like `functions`, `conditionals`, objects from other languages.

**Proposal 1** - rely on text decoration like bolding and underlining
- _works great for some themes, a good option in some cases._
- _insufficient for certain themes {low saturation, bright text etc.}_

**Proposal 2** - use existing surface colors as text colors in these special cases
- _we experimented with the color combinations that appeared promising and found the combinations to be poor. {contrast test failures, poor readability, crappy aesthetics}_
- _this method also feels inconsistent... using surface colors for text... YES it happens in some other places like buttons in certain states... we risk some really bad looking code editing situations._

**Proposal 3** -- add some more `text-colors`
- the downside: more objects to manage.
- the upside: more user freedom when creating custom themes (or editing existing themes)
- more colooooors! 🍎🍏💙💚💛💜📗📘



We settled on proposal 3. 
experiments are ongoing.

**Comment by lansingthomas** [comment]
Would these groups be acceptable?
1. literals (things that eval to themselves)
2. macros (incl. functions)special forms 
3. keyword arguments (symbols in the keyword package)
4. ;comments

**Comment by aartaka** [comment]
>1. literals (things that eval to themselves)

Yes.

>2. macros (incl. functions)special forms 

Not sure I get the idea, but I'm inclined to have macros+specials as one group, and functions as another.

>3. keyword arguments (symbols in the keyword package)

Might (should?) be merged with literals (although my Emacs theme colors them differently. Fancy and useless 😃)

>4. ;comments

Yes.

So , to summarize my idea, it's four groups: (1) regular code/symbols/functions/variables/etc. (you haven't mentioned this one, is it implied? I decided to include it for completeness); (2) literals+keywords (3) macros+specials (4) comments.

> (23 variables) x (10 modifiers)

Even thinking about that fills me with horror.

**Comment by aartaka** [comment]
>aartaka I know we talked about accommodating other languages in the structure-

Actually, our 4 semantic code groups (see comment above) apply to most other programming languages, and they might be a nice minimalist theme for them. I mean, when we're done with code colors, I might even use the idea in my Emacs coding setup 😂

**Comment by lansingthomas** [comment]
> Even thinking about that fills me with horror.

yes run! Flee while you can.

> (1) regular code/symbols/functions/variables/etc. (you haven't mentioned this one, is it implied?...)

yes, `text` covers this for now

**Comment by lansingthomas** [comment]
I think having `literals` and `keyword-arguments` as one item makes sense

**Comment by lansingthomas** [comment]
I keep running up against the limit of my knowledge here.

@jmercouris 
Can you help us group this dumb word salad into a coherent structure?

**Comment by lansingthomas** [comment]
light theme mockup: NOT the final colors
![testing codeblock themes (Light) - proposed A](https://github.com/atlas-engineer/nyxt/assets/6666079/c1210140-6c30-4a12-892e-3eca67eef0ad)

thoughts?

**Comment by lansingthomas** [comment]
![testing codeblock themes (Dark Amphora) - proposal](https://github.com/atlas-engineer/nyxt/assets/6666079/8cf9316d-aa08-43f9-919f-81dce456021f)

**Comment by aartaka** [comment]
Looks like a good grouping, I like it!
